core/src/test/java/com/linecorp/armeria/common/stream/StreamMessageOutputStreamTest.java

```diff
@@ -0,0 +1,343 @@
+                writer, x -> Integer.valueOf(x.byteBuf().readByte()));
+
+        assertDoesNotThrow(outputStream::close);
+        StepVerifier.create(writer)
+                    .expectNext(1, 2, 3, 4, 5)
+                    .verifyComplete();
+
+        for (int i = 0; i < 10; i++) {
+            assertDoesNotThrow(outputStream::close);
+            assertThatThrownBy(() -> outputStream.write(0))
+                    .isInstanceOf(IOException.class)
+                    .hasMessage("Stream closed");
+            assertThatThrownBy(() -> outputStream.write(new byte[] { 0 }, 0, 1))
+                    .isInstanceOf(IOException.class)
+                    .hasMessage("Stream closed");
+            assertThatThrownBy(outputStream::flush)
+                    .isInstanceOf(IOException.class)
+                    .hasMessage("Stream closed");
+        }
+    }
+
+    @Test
+    void flush() throws IOException {
+        final StreamMessage<Integer> source = StreamMessage.of(1, 2, 3, 4, 5);
+        final DefaultStreamMessage<Integer> writer = new DefaultStreamMessage<>();
+        final OutputStream outputStream = source.toOutputStream(
+                writer, x -> Integer.valueOf(x.byteBuf().readByte()));
+
+        final List<Integer> integers = ImmutableList.of(6, 7, 8, 9, 10);
+        for (Integer integer : integers) {
+            outputStream.write(integer.byteValue());
+            outputStream.flush();
+        }
+        writer.close();
+
+        StepVerifier.create(writer)
+                    .expectNext(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)
+                    .verifyComplete();
+    }
+
+    @Test
+    void streamMessage_aborted() throws IOException {
+        final Publisher<Integer> publisher = Flux.range(1, 10);
+        final StreamMessage<Integer> streamMessage = new PublisherBasedStreamMessage<>(publisher);
+        final StreamMessage<Integer> aborted = streamMessage
+                .peek(x -> {
+                    if (x == 6) {
+                        streamMessage.abort();
+                    }
+                });
+        final DefaultStreamMessage<Integer> writer = new DefaultStreamMessage<>();
+        final OutputStream outputStream = aborted.toOutputStream(
+                writer, x -> Integer.valueOf(x.byteBuf().readByte()));
+
+        outputStream.write(11);
+        assertThatThrownBy(outputStream::flush)
+                .isInstanceOf(CompletionException.class)
+                .hasCauseInstanceOf(AbortedStreamException.class);
+        assertThatThrownBy(() -> outputStream.write(0))
+                .isInstanceOf(IOException.class)
+                .hasMessage("Stream closed");
+        assertThatThrownBy(outputStream::flush).isInstanceOf(IOException.class)
+                                               .hasMessage("Stream closed");
+        assertDoesNotThrow(outputStream::close);
+        assertThatThrownBy(() -> writer.write(0)).isInstanceOf(ClosedStreamException.class);
```

Comment:
(Question) If StreamMessage is aborted, when should we detect this?
1. on `write()` (subscribe StreamMessage on write())
2. on `flush()` (subscribe StreamMessage lazily on flush())

Also, I'm not sure about proper exception type here 😅
- If StreamMessage is aborted, should we throw `AbortedStreamException` or `IOException`?
- When aborted, then should we close `OutputStream` & `DefaultStreamMessage`?

I have lots of questions here so feel free to give me any suggestion!